android/src/main/java/com/robinpowered/RNFrequency/RNFrequencyModule.java

```diff
@@ -0,0 +1,116 @@
+
+  public RNFrequencyModule(ReactApplicationContext reactContext) {
+    super(reactContext);
+
+    final ReactApplicationContext thisContext = reactContext;
+
+    IntentFilter headphonesFilter = new IntentFilter(Intent.ACTION_HEADSET_PLUG);
+    BroadcastReceiver headphonesReceiver = new BroadcastReceiver() {
+        @Override
+        public void onReceive(Context context, Intent intent) {
+            boolean headsetPluggedIn = false;
+            int state = intent.getIntExtra("state", -1);
+            switch (state) {
+                case (0):
+                    headsetPluggedIn = false;
+                    break;
+                case (1):
+                    headsetPluggedIn = true;
+                    break;
+                default:
+                    headsetPluggedIn = false;
+            }
+            WritableNativeMap data = new WritableNativeMap();
+            data.putBoolean("audioJackPluggedIn", headsetPluggedIn);
+            if (thisContext.hasActiveCatalystInstance()) {
+                thisContext.getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter.class).emit(AUDIO_CHANGED_NOTIFICATION,
+                        data);
+            }
+        }
+    };
+    thisContext.registerReceiver(headphonesReceiver, headphonesFilter);
+  }
+
+  private boolean isHeadSetPluggedIn() {
+    AudioManager am = (AudioManager) getReactApplicationContext().getSystemService(Context.AUDIO_SERVICE);
+    if (Build.VERSION.SDK_INT < Build.VERSION_CODES.M) {
+      return am.isWiredHeadsetOn();
+    } else {
+      AudioDeviceInfo[] devices = am.getDevices(AudioManager.GET_DEVICES_OUTPUTS);
+      for (int i = 0; i < devices.length; i++) {
+        AudioDeviceInfo device = devices[i];
+        if (device.getType() == AudioDeviceInfo.TYPE_WIRED_HEADPHONES) {
+          return true;
+        }
+      }
+    }
+    return false;
+  }
+
+  @Override
+  public @Nullable Map<String, Object> getConstants() {
+    HashMap<String, Object> constants = new HashMap<String, Object>();
+    constants.put(AUDIO_CHANGED_NOTIFICATION, AUDIO_CHANGED_NOTIFICATION);
+    return constants;
+  }
+
+  @Override
+  public String getName() {
+    return moduleName;
+  }
+
+  @ReactMethod
+  public void playFrequency(double frequency, double duration, final Promise promise) {
+    final int dur = (int) duration;
+    final int freq = (int) frequency;
+
+    int count = (int)(44100.0 * 2.0 * (dur)) & ~1;
+
+    short[] samples = new short[count];
+    for(int i = 0; i < count; i += 2){
+      short sample = (short)(Math.sin(2 * Math.PI * i / (44100.0 / freq)) * 0x7FFF);
+      samples[i + 0] = sample;
+      samples[i + 1] = sample;
+    }
+
+    AudioTrack track = new AudioTrack(AudioManager.STREAM_MUSIC, 44100,
+      AudioFormat.CHANNEL_OUT_STEREO, AudioFormat.ENCODING_PCM_16BIT,
+      count * (Short.SIZE / 8), AudioTrack.MODE_STATIC);
+
+    track.write(samples, 0, count);
+    track.play();
+
+    promise.resolve(true);
```

Comment:
Does the `track` offer the ability to notify us when it's completed? If it does... wondering if we should resolve the promise at that point in time

Reply:
Good point, I'll look into that.

Reply:
I ask because there could be a day where we want to sequence a few colors (who knows?!), and it would be much nicer to be able to do:


```js
Frequency.play(freqA, durationA)
  .then(() => Frequency.play(freqB, durationB))
  .then(() => Frequency.play(freqC, durationC))
```

Than
```js
Frequency.play(freqA, durationA)
setTimeout(() => {
  Frequency.play(freqB, durationB)
  setTimeout(() => {
    Frequency.play(freqC, durationC)
  }, durationB)
}, durationA)
```

Reply:
Noted. Seems like there's indeed a way to send out a notification after playback so I'll be implementing that. 😄